src/NUnitEngine/nunit.engine/Services/ExtensionService.cs

```diff
@@ -0,0 +1,154 @@
+{
+    public class ExtensionService : Service
+    {
+        private List<ExtensionPoint> _extensionPoints = new List<ExtensionPoint>();
+        private Dictionary<string, ExtensionPoint> _pathIndex = new Dictionary<string, ExtensionPoint>();
+        private Dictionary<Type, ExtensionPoint> _typeIndex = new Dictionary<Type, ExtensionPoint>();
+
+        private List<ExtensionNode> _extensions = new List<ExtensionNode>();
+
+        #region Public Properties
+
+        public IEnumerable<ExtensionPoint> ExtensionPoints
+        {
+            get { return _extensionPoints; }
+        }
+
+        public IEnumerable<ExtensionNode> Extensions
+        {
+            get { return _extensions;  }
+        }
+
+        #endregion
+
+        public ExtensionPoint GetExtensionPoint(string path)
+        {
+            return _pathIndex[path];
+        }
+
+        public ExtensionPoint GetExtensionPoint(Type type)
+        {
+            return _typeIndex[type];
+        }
+
+        public T[] GetExtensions<T>()
+        {
+            var extensions = new List<T>();
+
+            var ep = GetExtensionPoint(typeof(T));
+            foreach (var node in ep.Extensions)
+                extensions.Add((T)node.ExtensionObject);
+
+            return extensions.ToArray();
+        }
+
+        #region Service Overrides
+
+        public override void StartService()
+        {
+            var thisAssembly = Assembly.GetExecutingAssembly();
+            var startDir = new DirectoryInfo(AssemblyHelper.GetDirectoryName(thisAssembly));
+
+            FindExtensionPoints(thisAssembly);
+            FindExtensionsInDirectory(startDir);
+
+            base.StartService();
+        }
+
+        private void FindExtensionPoints(Assembly assembly)
+        {
+            foreach (ExtensionPointAttribute attr in assembly.GetCustomAttributes(typeof(ExtensionPointAttribute), false))
+            {
+                var ep = new ExtensionPoint(attr.Path, attr.Type)
+                {
+                    Description = attr.Description
+                };
+
+                _extensionPoints.Add(ep);
+
+                // TODO: These Adds will throw if the path or type is a duplicate.
+                // We should give an error message instead.
+                _pathIndex.Add(ep.Path, ep);
+                _typeIndex.Add(ep.Type, ep);
+            }
+        }
+
+        private void FindExtensionsInDirectory(DirectoryInfo startDir)
+        {
+            var addinsFiles = startDir.GetFiles("*.addins");
+            if (addinsFiles.Length > 0)
+                foreach (var file in addinsFiles)
+                    ProcessAddinsFile(startDir.FullName, file.FullName);
+            else
+                foreach (var file in startDir.GetFiles("*.dll"))
+                    FindExtensionsInAssembly(file.FullName);
+        }
+
+        private void ProcessAddinsFile(string baseDir, string fileName)
+        {
+            var doc = new XmlDocument();
```

Comment:
There is YamlDotNet, http://aaubry.net/pages/yamldotnet.html, but it is fairly heavy for what we need. We could probably write a simple one that satisfies the subset of YAML we need;

1. Split lines at #, ignore second part (comments)
2. Ignore blank or empty lines
3. Lines starting with no whitespace and ending in :, switch state to adding to include, exclude, etc.
4. Lines starting with whitespace, trim whitespace and add to include, exclude, etc depending on state.
5. Anything else is an error

This will only allow for labels and items, but will probably do what we need?